### PR TITLE
Add payment mode selection for organisers

### PR DESCRIPTION
**Pull request by rishabhs95**
add capability for organisers to select payment modes for their tickets as Online/Offline.

### CODE REVIEW

**Comment by hennevogel** [comment]
You really think people want to mix online/offline payment methods?

**Comment by rishabhs95** [comment]
i thought this will provide freedom to organiser to select which payment mode they will prefer for each ticket.

they can choose whether they want to receive payments for it online/offline.

**Comment by rishabhs95** [comment]
please suggest if it should be flexible with payment mode associated to each tickets or on a larger level(like conference)?

File: db/migrate/20160526142456_add_payment_mode_to_tickets.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddPaymentModeToTickets < ActiveRecord::Migration
+  def change
+    add_column :tickets, :payment_mode, :string
```
Comment (lagartoflojo) [inline db/migrate/20160526142456_add_payment_mode_to_tickets.rb]: By adding the column like this, all old tickets will have a NULL `payment_method`. Is that valid? Should we do something about that?

File: app/models/ticket.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ class Ticket < ActiveRecord::Base
   # If we would allow different currencies per conference we also have to handle convertions between currencies!
   validate :tickets_of_conference_have_same_currency
 
-  validates :price_cents, :price_currency, :title, presence: true
+  validates :price_cents, :price_currency, :payment_mode, :title, presence: true
```
Comment (ChrisBr) [inline app/models/ticket.rb]: Please also add an inclusion validation as I see you only allow the values Online/Offline in the view.
http://guides.rubyonrails.org/active_record_validations.html#inclusion

File: spec/models/ticket_spec.rb [hunk 3]
```diff
@@ -22,6 +22,10 @@
       should validate_presence_of(:price_currency)
     end
 
+    it 'is not valid without a payment_mode' do
```
Comment (ChrisBr) [inline spec/models/ticket_spec.rb]: Also add a test for the inclusion validation

File: db/migrate/20160526142456_add_payment_mode_to_tickets.rb (hunk 1 above)
Comment (rishabhs95) [inline db/migrate/20160526142456_add_payment_mode_to_tickets.rb]: yes... i will edit it keeping "Offline" as default payment_mode as the existing tickets required offline payments

File: spec/models/ticket_spec.rb (hunk 3 above)
Comment (rishabhs95) [inline spec/models/ticket_spec.rb]: ok :)

**Comment by rishabhs95** [comment]
Just a clarification for the use case for associating payment_mode with tickets: 

According to my understanding, a conference is an overall event and various talks and other events are organised inside it. These events can be of two types: paid and free. The payment method for the paid events can be further classified into online and offline payment.

One example of event which requires online payment would be provision of accommodation or some other special paid talks. Online payment signifies commitment on attendees behalf and the organisers can focus on arrangements instead of last minute 'no-shows'.

However, there are other events like tours or parties which attendees are encouraged to decide on the spot. The organisers require a rough estimate of number of attendees. Offline payments is suitable in such scenario.

**Comment by hennevogel** [comment]
@rishabhs95 tickets are for a conference as a whole. Not for individual happenings during it

**Comment by rishabhs95** [comment]
Okay @hennevogel, I will make the changes accordingly

**Comment by hennevogel** [comment]
Please send a new pull request then :-)

**Pull request closed without merging**